setup/gcp/installer/util/merge_engine.py

```diff
@@ -0,0 +1,259 @@
+        identifier = field_identifiers.get('default_identifier')
+        merge_dict_list(merge_from, merge_to, identifier,
+                        fields_to_ignore, field_identifiers)
+    elif isinstance(merge_to, dict) and isinstance(merge_from, dict):
+        merge_dict(merge_from, merge_to, fields_to_ignore, field_identifiers)
+    else:
+        raise FormatNotSupported(
+            'Merging {} and {} is not supported'.format(
+                type(merge_to), type(merge_from)))
+
+
+def merge_dict(merge_from_dict, merge_to_dict,
+               fields_to_ignore, field_identifiers):
+    """Merge target_dict into base_dict.
+
+    Note: merge_to_dict will be modified during the merge process.
+
+    Args:
+        merge_from_dict (dict): Dictionary merging from.
+        merge_to_dict (dict): Dictionary merging to.
+        fields_to_ignore (list): Fields to ignore (keep in base_dict).
+        field_identifiers (dict): Identifiers for fields.
+    """
+
+    for key, val in merge_to_dict.iteritems():
+        if key in merge_from_dict:
+            # If merge_from_dict has the same key, we check if the value is
+            # an instance of dictionary. If it is we merge recursively and
+            # if it's not, we do a simple replace if the key is not in
+            # fields_to_ignore.
+            if key in fields_to_ignore:
+                continue
+            if isinstance(val, dict):
+                merge_dict(merge_from_dict.get(key), val,
+                           fields_to_ignore, field_identifiers)
+            elif isinstance(val, list) and val and isinstance(val[0], dict):
+                # If the list has at least one item, we check for the type
+                # of the first item, if it's a dictionary, we invoke
+                # merge_dict_list.
+                identifier = field_identifiers.get(key)
+                if isinstance(identifier, list):
+                    identifier = _select_identifier(identifier, val[0])
+                merge_dict_list(merge_from_dict.get(key), val, identifier,
+                                fields_to_ignore, field_identifiers)
+            else:
+                merge_to_dict[key] = merge_from_dict.get(key)
+
+    for key, val in merge_from_dict.iteritems():
+        if key not in merge_to_dict and key not in fields_to_ignore:
+            # If this is a key we have only in target but not in base, we add
+            # it to the base_dict.
+            merge_to_dict[key] = val
+
+
+def _select_identifier(identifiers, dict_to_identify):
+    """Select the right field identifier.
+
+    Args:
+        identifiers (list): List of identifiers.
+        dict_to_identify (dict): Dictionary to identify.
+
+    Returns:
+        str: Identifier.
+    """
+
+    for identifier in identifiers:
+        if identifier in dict_to_identify:
+            return identifier
+    return ''
+
+
+def merge_dict_list(merge_from_dict_list, merge_to_dict_list, identifier,
```

Comment:
The combination of `dict_list` seems can be improved.  How about this?  The plural forms is how lists are referenced in python.
```
def merge_dicts(merge_from_dicts, merge_to_dicts,
```

Reply:
Updated